server/middleware/reactApplication/index.js

```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import Helmet from 'react-helmet';
+import { renderToString, renderToStaticMarkup } from 'react-dom/server';
+import { StaticRouter } from 'react-router-dom';
+import { Provider } from 'react-redux';
+import { runJobs } from 'react-jobs/ssr';
+import configureStore from '../../../shared/redux/configureStore';
+
+import config from '../../../config';
+
+import ServerHTML from './ServerHTML';
+import DemoApp from '../../../shared/components/DemoApp';
+
+/**
+ * React application middleware, supports server side rendering.
+ */
+export default function reactApplicationMiddleware(request, response) {
+  // Ensure a nonce has been provided to us.
+  // See the server/middleware/security.js for more info.
+  if (typeof response.locals.nonce !== 'string') {
+    throw new Error('A "nonce" value has not been attached to the response');
+  }
+  const nonce = response.locals.nonce;
+
+  // It's possible to disable SSR, which can be useful in development mode.
+  // In this case traditional client side only rendering will occur.
+  if (config('disableSSR')) {
+    if (process.env.BUILD_FLAG_IS_DEV) {
+      // eslint-disable-next-line no-console
+      console.log('==> Handling react route without SSR');
+    }
+    // SSR is disabled so we will return an "empty" html page and
+    // rely on the client to initialize and render the react application.
+    const html = renderToStaticMarkup(<ServerHTML nonce={nonce} />);
+    response.status(200).send(html);
+    return;
+  }
+
+  // Create the redux store.
+  const store = configureStore();
+  const { getState } = store;
+
+  // First create a context for <StaticRouter>, which will allow us to
+  // query for the results of the render.
+  const reactRouterContext = {};
+
+  // Declare our React application.
+  const app = (
+    <StaticRouter location={request.url} context={reactRouterContext}>
+      <Provider store={store}>
+        <DemoApp />
+      </Provider>
+    </StaticRouter>
+  );
+
+  // Pass our app into the react-async-component helper so that any async
+  // components are resolved for the render.
+  runJobs(app).then(({ appWithAsyncComponents, state, STATE_IDENTIFIER }) => {
+    // Generate the html response.
+    const html = renderToStaticMarkup(
+      <ServerHTML
+        reactAppString={renderToString(appWithAsyncComponents)}
+        nonce={nonce}
+        helmet={Helmet.rewind()}
+        asyncComponents={{ state, STATE_IDENTIFIER }}
+        initialState={getState()}
+        jobsTate={{ state, STATE_IDENTIFIER }}
```

Comment:
@Neophy7e hey man, i'm not sure about this details, i need some help from @ctrlplusb to understand this.